src/Microsoft.SqlTools.Hosting.v2/Utility/Logger.cs

```diff
@@ -5,86 +5,152 @@
+    public enum LogEvent : ushort
     {
-        /// <summary>
-        /// Indicates a verbose log message.
-        /// </summary>
-        Verbose,
+        Default = 0,
+        IoFileSystem = 1,
+        OsSubSystem = 2,
+    }
+
+    /// <summary>
+    /// Provides a simple logging interface built on top of .Net tracing frameworks
+    /// </summary>
+    public static class Logger
+    {
+        internal const SourceLevels defaultTracingLevel = SourceLevels.Critical; 
+        internal const string defaultTraceSrouce = "sqltools";
+        private static SourceLevels tracingLevel = defaultTracingLevel;
+        private static string logFileFullPath;
+
+        internal static TraceSource TraceSource { get; set; }
+        internal static string LogFileFullPath
+        {
+            get => logFileFullPath;
+            private set
+            {
+                //If the log file path has a directory component then ensure that the directory exists.
+                if (!string.IsNullOrEmpty(Path.GetDirectoryName(value)) && !Directory.Exists(Path.GetDirectoryName(value)))
+                    Directory.CreateDirectory(Path.GetDirectoryName(value));
+                logFileFullPath = value;
+                ConfigureListener();
+            }
+        }
+
+        private static SqlToolsTraceListener Listener { get; set; }
+
+        private static void ConfigureLogFile(string logFilePrefix)
+        {
+            LogFileFullPath = GenerateLogFilePath(logFilePrefix);
+        }
 
         /// <summary>
-        /// Indicates a normal, non-verbose log message.
+        /// Calling method will turn on inclusion CallStack in the log for all future traces
         /// </summary>
-        Normal,
+        public static void StartCallStack() => Listener.TraceOutputOptions |= TraceOptions.Callstack;
 
         /// <summary>
-        /// Indicates a warning message.
+        /// Calling method will turn off inclusion of CallStack in the log for all future traces
         /// </summary>
-        Warning,
+        public static void StopCallStack() => Listener.TraceOutputOptions &= ~TraceOptions.Callstack;
 
         /// <summary>
-        /// Indicates an error message.
+        /// Calls flush on defaultTracingLevel configured listeners.
         /// </summary>
-        Error
-    }
-
-    /// <summary>
-    /// Provides a simple logging interface.  May be replaced with a
-    /// more robust solution at a later date.
-    /// </summary>
-    public class Logger
-    {
-        private static LogWriter logWriter;
-
-        private static bool isEnabled;
-
-        private static bool isInitialized = false;
-
-        private static Lazy<Logger> lazyInstance = new Lazy<Logger>(() => new Logger());
+        public static void Flush()
+        {
+            TraceSource.Flush();
+            Trace.Flush();
+        }
 
-        public static Logger Instance
+        public static void Close()
+        {
+            Flush();
```

Comment:
Can these ever error? If so do we expect to null out the LIstener anyhow?

Reply:
As per msdn the close does not throw, so error is not expected.
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.trace.close?view=netcore-2.1